libraries/AP_UAVCAN/AP_UAVCAN.cpp

```diff
@@ -319,6 +322,80 @@ void AP_UAVCAN::init(uint8_t driver_index, bool enable_filters)
     debug_uavcan(2, "UAVCAN: init done\n\r");
 }
 
+// send ESC telemetry messages over MAVLink
+void AP_UAVCAN::send_esc_telemetry_mavlink(uint8_t mav_chan)
+{
+
+    WITH_SEMAPHORE(_esc_sem);
+
+    bool group[3] = {false};
+
+    // each ESC Mavlink message contain group of 4 ESC.
+    // find what groups have at least one ESC available data,
+    // and mark those groups for sending. 
+    for (uint8_t i = 0; i < UAVCAN_MAX_MAV_ESC_GROUPS; i++) { 
+        for (uint8_t j = 0; j < UAVCAN_MAV_ESC_PER_GROUP; j++) {
+            if (_escs_data[(i*4)+j].available) {
+                group[i] = true; 
+                break;
+            }
+        }
+    }
+
+    for (uint8_t i = 0; i < UAVCAN_MAX_MAV_ESC_GROUPS; i++) {
+
+        // skip this group of ESCs if no data available to send
+        if ( ! group[i] ) {
+            continue;
+        }
+
+        // arrays to hold output
+        uint8_t temperature[UAVCAN_MAV_ESC_PER_GROUP] {};
+        uint16_t voltage[UAVCAN_MAV_ESC_PER_GROUP] {};
+        uint16_t current[UAVCAN_MAV_ESC_PER_GROUP] {};
+        uint16_t current_tot[UAVCAN_MAV_ESC_PER_GROUP] {};
+        uint16_t rpm[UAVCAN_MAV_ESC_PER_GROUP] {};
+        uint16_t count[UAVCAN_MAV_ESC_PER_GROUP] {};
+
+        // fill in output arrays
+        for (uint8_t j = 0; j < UAVCAN_MAV_ESC_PER_GROUP; j++) {
+            uint8_t esc_id = i * UAVCAN_MAV_ESC_PER_GROUP + j;
+            if ( ! _escs_data[esc_id].available ) {
+                continue;
+            } 
+
+            _escs_data[esc_id].available = false;
+
+            temperature[j] = _escs_data[esc_id].temp;
+            voltage[j] = _escs_data[esc_id].voltage;
+            current[j] = _escs_data[esc_id].current;
+            current_tot[j] = 0; // currently not implemented
+            rpm[j] = _escs_data[esc_id].rpm;
+            count[j] = _escs_data[esc_id].count;
+        }
+
+        if (!HAVE_PAYLOAD_SPACE((mavlink_channel_t) mav_chan, ESC_TELEMETRY_1_TO_4)) {
```

Comment:
This is the way it's done in other drivers. Is this condition might change during code execution? If so this should be checked as late as possible because it might change just before sending the message. Don't you think?